mlflow/models/evaluation/base.py

```diff
@@ -425,9 +425,50 @@ def _hash_uint64_ndarray_as_bytes(array):
     return struct.pack(f">{array.size}Q", *array)
 
 
+def _is_empty_list_or_array(data):
+    if isinstance(data, list):
+        return len(data) == 0
+    elif isinstance(data, np.ndarray):
+        return data.size == 0 or data.ndim == 0
+    return False
+
+
+def _is_array_has_dict(nd_array):
+    if _is_empty_list_or_array(nd_array):
+        return False
+
+    # It is less likely the array or list contains heterogeneous elements, so just checking the
+    # first element to avoid performance overhead.
+    elm = nd_array[0]
+    if isinstance(elm, (list, np.ndarray)):
+        return _is_array_has_dict(elm)
+    elif isinstance(elm, dict):
+        return True
+
+    return False
+
+
+def _hash_array_of_dict_as_bytes(data):
+    # NB: If an array or list contains dictionary element, it can't be hashed with
+    # pandas.util.hash_array. Hence we need to manually hash the elements here. This is
+    # particularly for the LLM use case where the input can be a list of dictionary
+    # (chat/completion payloads), so doesn't handle more complex case like nested lists.
+    result = b""
+    for elm in data:
+        if isinstance(elm, (list, np.ndarray)):
+            result += _hash_array_of_dict_as_bytes(elm)
+        elif isinstance(elm, dict):
+            result += _hash_dict_as_bytes(elm)
```

Comment:
Can we move all of these hashing methods into a separate module, ideally located in `mlflow.data`? since these are all about hashing for use with MLflow Datasets

Reply:
Agree this should be in a separate module, but for the location, I would still put this under `evaluation` module. These methods are used for hashing various input like list, numpy, dataframe, and only used for `class EvaluationDataset` digest.

Reply:
I would prefer to put all of `EvaluationDataset` in `mlflow/data` to maintain consistency with other datasets - e.g. we don't put `TensorflowDataset` in `mlflow/tensorflow`

Reply:
Feel free to tackle this in a follow-up ticket as long as we track it :)

Reply:
> I would prefer to put all of EvaluationDataset in mlflow/data

Makes sense.

> Feel free to tackle this in a follow-up ticket as long as we track it :)

Yeah let me do that, created JIRA in next sprint (ML-39039), but may be able to pick up if no fun surprise happens in other tasks:)